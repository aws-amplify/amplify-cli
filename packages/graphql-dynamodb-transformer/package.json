{
  "name": "graphql-dynamodb-transformer",
  "version": "6.19.5",
  "description": "An AppSync model transform that takes a simple model and creates a DynamoDB table, DynamoDB stream.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/graphql-dynamodb-transformer"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "clean": "rimraf ./lib",
    "watch": "tsc -w"
  },
  "dependencies": {
    "@types/pluralize": "^0.0.29",
    "cloudform-types": "^4.2.0",
    "graphql": "^14.5.8",
    "graphql-mapping-template": "4.13.4",
    "graphql-transformer-common": "4.17.4",
    "graphql-transformer-core": "6.19.4",
    "md5": "^2.2.1",
    "pluralize": "^8.0.0"
  },
  "devDependencies": {
    "@types/md5": "^2.1.33",
    "@types/node": "^10.17.13"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*.test.ts)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
