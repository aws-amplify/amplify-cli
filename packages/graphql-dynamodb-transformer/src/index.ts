export * from './DynamoDBModelTransformer';
export * from './definitions';
export * from './ModelDirectiveArgs';
