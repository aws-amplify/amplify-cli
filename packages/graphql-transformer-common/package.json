{
  "name": "graphql-transformer-common",
  "version": "1.0.28",
  "description": "Common code and constants for AppSync Transformers",
  "main": "lib/index.js",
  "scripts": {
    "test": "jest",
    "build": "tsc",
    "clean": "rm -rf ./lib"
  },
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "dependencies": {
    "cloudform": "^2.2.1",
    "graphql": "^0.13.2",
    "graphql-mapping-template": "^1.0.28",
    "graphql-transformer-core": "^1.0.28"
  },
  "devDependencies": {
    "@types/graphql": "^0.13.1",
    "@types/jest": "^23.1.1",
    "@types/node": "^10.3.4",
    "aws-sdk": "^2.259.1",
    "jest": "^23.1.0",
    "ts-jest": "^22.4.6",
    "tslint": "^5.10.0",
    "typescript": "^2.8.3"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*.test.*)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ]
  }
}
