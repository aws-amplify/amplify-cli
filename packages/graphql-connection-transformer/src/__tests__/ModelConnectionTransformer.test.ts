import {
  ObjectTypeDefinitionNode,
  parse,
  FieldDefinitionNode,
  DocumentNode,
  DefinitionNode,
  Kind,
  InputObjectTypeDefinitionNode,
  InputValueDefinitionNode,
} from 'graphql';
import { GraphQLTransform, TRANSFORM_CURRENT_VERSION } from 'graphql-transformer-core';
import { ResolverResourceIDs, ModelResourceIDs, ResourceConstants } from 'graphql-transformer-common';
import { DynamoDBModelTransformer } from 'graphql-dynamodb-transformer';
import { ModelConnectionTransformer } from '../ModelConnectionTransformer';

test('Test ModelConnectionTransformer simple one to many happy case', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection
    }
    type Comment @model {
        id: ID!
        content: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();
  const schemaDoc = parse(out.schema);

  // Post.comments field
  const postType = getObjectType(schemaDoc, 'Post');
  expectFields(postType, ['comments']);
  const commentField = postType.fields.find(f => f.name.value === 'comments');
  expect(commentField.arguments.length).toEqual(4);
  expectArguments(commentField, ['filter', 'limit', 'nextToken', 'sortDirection']);
  expect(commentField.type.kind).toEqual(Kind.NAMED_TYPE);
  expect((commentField.type as any).name.value).toEqual('ModelCommentConnection');

  // Check the Comment.commentPostId
  // Check the Comment.commentPostId inputs
  const commentCreateInput = getInputType(schemaDoc, ModelResourceIDs.ModelCreateInputObjectName('Comment'));
  const connectionId = commentCreateInput.fields.find(f => f.name.value === 'postCommentsId');
  expect(connectionId).toBeTruthy();

  const commentUpdateInput = getInputType(schemaDoc, ModelResourceIDs.ModelUpdateInputObjectName('Comment'));
  const connectionUpdateId = commentUpdateInput.fields.find(f => f.name.value === 'postCommentsId');
  expect(connectionUpdateId).toBeTruthy();
});

test('Test ModelConnectionTransformer simple one to many happy case with custom keyField', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(keyField: "postId")
    }
    type Comment @model {
        id: ID!
        content: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();
  const schemaDoc = parse(out.schema);

  // Post.comments field
  const postType = getObjectType(schemaDoc, 'Post');
  expectFields(postType, ['comments']);
  const commentField = postType.fields.find(f => f.name.value === 'comments');
  expect(commentField.arguments.length).toEqual(4);
  expectArguments(commentField, ['filter', 'limit', 'nextToken', 'sortDirection']);
  expect(commentField.type.kind).toEqual(Kind.NAMED_TYPE);
  expect((commentField.type as any).name.value).toEqual('ModelCommentConnection');

  // Check the Comment.commentPostId
  // Check the Comment.commentPostId inputs
  const commentCreateInput = getInputType(schemaDoc, ModelResourceIDs.ModelCreateInputObjectName('Comment'));
  const connectionId = commentCreateInput.fields.find(f => f.name.value === 'postId');
  expect(connectionId).toBeTruthy();

  const commentUpdateInput = getInputType(schemaDoc, ModelResourceIDs.ModelUpdateInputObjectName('Comment'));
  const connectionUpdateId = commentUpdateInput.fields.find(f => f.name.value === 'postId');
  expect(connectionUpdateId).toBeTruthy();
});

test('Test that ModelConnection Transformer throws error when the field in connection is not found in the related Type', () => {
  const invalidSchema = `
  type Post @model {
    name: String!
    teamID: ID!
    team: Team @connection(fields: ["teamID"])
  }

  type Team @model {
    name: [String!]!
  }
  `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  try {
    transformer.transform(invalidSchema);
    expect(true).toEqual(false);
  } catch (e) {
    console.log(e);
    expect(e).toBeTruthy();
    expect(e.name).toEqual('InvalidDirectiveError');
  }
});

test('Test ModelConnectionTransformer simple one to many happy case with custom keyField', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        createdAt: String
        updatedAt: String
        comments: [Comment] @connection(name: "PostComments", keyField: "postId")
    }
    type Comment @model {
        id: ID!
        content: String!
        post: Post! @connection(name: "PostComments", keyField: "postId")
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();
  const schemaDoc = parse(out.schema);

  // Post.comments field
  const postType = getObjectType(schemaDoc, 'Post');
  expectFields(postType, ['comments']);
  const commentField = postType.fields.find(f => f.name.value === 'comments');
  expect(commentField.arguments.length).toEqual(4);
  expectArguments(commentField, ['filter', 'limit', 'nextToken', 'sortDirection']);
  expect(commentField.type.kind).toEqual(Kind.NAMED_TYPE);
  expect((commentField.type as any).name.value).toEqual('ModelCommentConnection');

  // Check the Comment.commentPostId
  // Check the Comment.commentPostId inputs
  const commentCreateInput = getInputType(schemaDoc, ModelResourceIDs.ModelCreateInputObjectName('Comment'));
  const connectionId = commentCreateInput.fields.find(f => f.name.value === 'postId');
  expect(connectionId).toBeTruthy();
  expect(connectionId.type.kind).toEqual(Kind.NON_NULL_TYPE);

  const commentUpdateInput = getInputType(schemaDoc, ModelResourceIDs.ModelUpdateInputObjectName('Comment'));
  const connectionUpdateId = commentUpdateInput.fields.find(f => f.name.value === 'postId');
  expect(connectionUpdateId).toBeTruthy();
  expect(connectionUpdateId.type.kind).toEqual(Kind.NAMED_TYPE);
});

test('Test ModelConnectionTransformer complex one to many happy case', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "PostComments")
    }
    type Comment @model {
        id: ID!
        content: String
        post: Post @connection(name: "PostComments")
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Comment', 'post')]).toBeTruthy();
  const schemaDoc = parse(out.schema);
  const postType = getObjectType(schemaDoc, 'Post');
  const commentType = getObjectType(schemaDoc, 'Comment');

  // Check Post.comments field
  expectFields(postType, ['comments']);
  const commentField = postType.fields.find(f => f.name.value === 'comments');
  expect(commentField.arguments.length).toEqual(4);
  expectArguments(commentField, ['filter', 'limit', 'nextToken', 'sortDirection']);
  expect(commentField.type.kind).toEqual(Kind.NAMED_TYPE);
  expect((commentField.type as any).name.value).toEqual('ModelCommentConnection');

  // Check the Comment.commentPostId inputs
  const commentCreateInput = getInputType(schemaDoc, ModelResourceIDs.ModelCreateInputObjectName('Comment'));
  const connectionId = commentCreateInput.fields.find(f => f.name.value === 'commentPostId');
  expect(connectionId).toBeTruthy();

  const commentUpdateInput = getInputType(schemaDoc, ModelResourceIDs.ModelUpdateInputObjectName('Comment'));
  const connectionUpdateId = commentUpdateInput.fields.find(f => f.name.value === 'commentPostId');
  expect(connectionUpdateId).toBeTruthy();

  // Check Comment.post field
  const postField = commentType.fields.find(f => f.name.value === 'post');
  expect(postField.arguments.length).toEqual(0);
  expect(postField.type.kind).toEqual(Kind.NAMED_TYPE);
  expect((postField.type as any).name.value).toEqual('Post');
});

test('Test ModelConnectionTransformer many to many should fail', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "ManyToMany")
    }
    type Comment @model {
        id: ID!
        content: String
        posts: [Post] @connection(name: "ManyToMany")
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  try {
    transformer.transform(validSchema);
    expect(true).toEqual(false);
  } catch (e) {
    // Should throw bc we don't let support many to many
    expect(e).toBeTruthy();
    expect(e.name).toEqual('InvalidDirectiveError');
  }
});

test('Test ModelConnectionTransformer many to many should fail due to missing other "name"', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "ManyToMany")
    }
    type Comment @model {
        id: ID!
        content: String

        # This is meant to be the other half of "ManyToMany" but I forgot.
        posts: [Post] @connection
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  try {
    transformer.transform(validSchema);
    expect(true).toEqual(false);
  } catch (e) {
    // Should throw bc we check both halves when name is given
    expect(e).toBeTruthy();
    expect(e.name).toEqual('InvalidDirectiveError');
  }
});

test('Test ModelConnectionTransformer many to many should fail due to missing other "name"', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        things: [Thing!] @connection
    }

    type Thing @model(queries: null, mutations: null) {
        id: ID!
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'things')]).toBeTruthy();
  const schemaDoc = parse(out.schema);
  const postType = getObjectType(schemaDoc, 'Post');
  const postConnection = getObjectType(schemaDoc, 'ModelPostConnection');
  const thingConnection = getObjectType(schemaDoc, 'ModelThingConnection');
  const thingFilterInput = getInputType(schemaDoc, 'ModelThingFilterInput');
  expect(thingFilterInput).toBeDefined();
  expect(postType).toBeDefined();
  expect(thingConnection).toBeDefined();
  expect(postConnection).toBeDefined();
});

test('Test ModelConnectionTransformer with non null @connections', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        createdAt: String
        updatedAt: String
        comments: [Comment] @connection(name: "PostComments", keyField: "postId")

        # A non null on the one in a 1-M does enforce a non-null
        # on the CreatePostInput
        singleComment: Comment! @connection

        # A non null on the many in a 1-M does not enforce a non-null
        # in the CommentCreateInput because it is not explicitly implied.
        manyComments: [Comment]! @connection
    }
    type Comment @model {
        id: ID!
        content: String!

        # A non-null on the one in 1-M again enforces a non null.
        post: Post! @connection(name: "PostComments", keyField: "postId")
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();
  const schemaDoc = parse(out.schema);

  // Post.comments field
  const postType = getObjectType(schemaDoc, 'Post');
  expectFields(postType, ['comments']);
  const commentField = postType.fields.find(f => f.name.value === 'comments');
  expect(commentField.arguments.length).toEqual(4);
  expectArguments(commentField, ['filter', 'limit', 'nextToken', 'sortDirection']);
  expect(commentField.type.kind).toEqual(Kind.NAMED_TYPE);
  expect((commentField.type as any).name.value).toEqual('ModelCommentConnection');

  // Check the Comment.commentPostId
  // Check the Comment.commentPostId inputs
  const commentCreateInput = getInputType(schemaDoc, ModelResourceIDs.ModelCreateInputObjectName('Comment'));
  const connectionId = commentCreateInput.fields.find(f => f.name.value === 'postId');
  expect(connectionId).toBeTruthy();
  expect(connectionId.type.kind).toEqual(Kind.NON_NULL_TYPE);

  const manyCommentId = commentCreateInput.fields.find(f => f.name.value === 'postManyCommentsId');
  expect(manyCommentId).toBeTruthy();
  expect(manyCommentId.type.kind).toEqual(Kind.NAMED_TYPE);

  const commentUpdateInput = getInputType(schemaDoc, ModelResourceIDs.ModelUpdateInputObjectName('Comment'));
  const connectionUpdateId = commentUpdateInput.fields.find(f => f.name.value === 'postId');
  expect(connectionUpdateId).toBeTruthy();
  expect(connectionUpdateId.type.kind).toEqual(Kind.NAMED_TYPE);

  // Check the post create type
  const postCreateInput = getInputType(schemaDoc, ModelResourceIDs.ModelCreateInputObjectName('Post'));
  const postConnectionId = postCreateInput.fields.find(f => f.name.value === 'postSingleCommentId');
  expect(postConnectionId).toBeTruthy();
  expect(postConnectionId.type.kind).toEqual(Kind.NON_NULL_TYPE);
});

test('Test ModelConnectionTransformer with sortField fails if not specified in associated type', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "PostComments", sortField: "createdAt")
    }
    type Comment @model {
        id: ID!
        content: String
        post: Post @connection(name: "PostComments")
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  expect(() => {
    transformer.transform(validSchema);
  }).toThrowError();
});

test('Test ModelConnectionTransformer with sortField creates a connection resolver with a sort key condition.', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "PostComments", sortField: "createdAt")
    }
    type Comment @model {
        id: ID!
        content: String
        post: Post @connection(name: "PostComments", sortField: "createdAt")
        createdAt: AWSDateTime
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();
  const schemaDoc = parse(out.schema);

  // Post.comments field
  const postType = getObjectType(schemaDoc, 'Post');
  expectFields(postType, ['comments']);
  const commentField = postType.fields.find(f => f.name.value === 'comments');
  expect(commentField.arguments.length).toEqual(5);
  expectArguments(commentField, ['createdAt', 'filter', 'limit', 'nextToken', 'sortDirection']);
});

test('Test ModelConnectionTransformer throws with invalid key fields', () => {
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });

  const invalidSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(keyField: "postId")
    }
    type Comment @model {
        id: ID!
        content: String

        # Key fields must be String or ID.
        postId: [String]
    }
    `;
  expect(() => transformer.transform(invalidSchema)).toThrow();

  const invalidSchema2 = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "PostComments", keyField: "postId")
    }
    type Comment @model {
        id: ID!
        content: String

        # Key fields must be String or ID.
        postId: [String]

        post: Post @connection(name: "PostComments", keyField: "postId")
    }
    `;
  expect(() => transformer.transform(invalidSchema2)).toThrow();

  const invalidSchema3 = `
    type Post @model {
        id: ID!
        title: String!
    }
    type Comment @model {
        id: ID!
        content: String

        # Key fields must be String or ID.
        postId: [String]

        post: Post @connection(keyField: "postId")
    }
    `;
  expect(() => transformer.transform(invalidSchema3)).toThrow();
});

test('Test ModelConnectionTransformer does not throw with valid key fields', () => {
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });

  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(keyField: "postId")
    }
    type Comment @model {
        id: ID!
        content: String

        # Key fields must be String or ID.
        postId: String
    }
    `;
  expect(() => transformer.transform(validSchema)).toBeTruthy();

  const validSchema2 = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(name: "PostComments", keyField: "postId")
    }
    type Comment @model {
        id: ID!
        content: String

        # Key fields must be String or ID.
        postId: ID

        post: Post @connection(name: "PostComments", keyField: "postId")
    }
    `;
  expect(() => transformer.transform(validSchema2)).toBeTruthy();

  const validSchema3 = `
    type Post @model {
        id: ID!
        title: String!
    }
    type Comment @model {
        id: ID!
        content: String

        # Key fields must be String or ID.
        postId: String

        post: Post @connection(keyField: "postId")
    }
    `;
  expect(() => transformer.transform(validSchema3)).toBeTruthy();
});

test('Test ModelConnectionTransformer sortField with missing @key should fail', () => {
  const validSchema = `
    type Model1 @model(subscriptions: null)
    {
        id: ID!
        sort: Int!
        name: String!
    }
    type Model2 @model(subscriptions: null)
    {
        id: ID!
        connection: Model1 @connection(sortField: "modelOneSort")
        modelOneSort: Int!
    }
        `;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });

  try {
    transformer.transform(validSchema);
    expect(true).toEqual(false);
  } catch (e) {
    expect(e).toBeTruthy();
    expect(e.name).toEqual('InvalidDirectiveError');
  }
});

test('Test ModelConnectionTransformer overrides the default limit', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(limit: 50)
    }
    type Comment @model {
        id: ID!
        content: String
    }
    `;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });

  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();

  // Post.comments field
  expect(out.resolvers['Post.comments.req.vtl']).toContain('#set( $limit = $util.defaultIfNull($context.args.limit, 50) )');
});

test('Test ModelConnectionTransformer uses the default limit', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection
    }
    type Comment @model {
        id: ID!
        content: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();

  // Post.comments field
  expect(out.resolvers['Post.comments.req.vtl']).toContain(
    `#set( $limit = $util.defaultIfNull($context.args.limit, ${ResourceConstants.DEFAULT_PAGE_LIMIT}) )`,
  );
});

test('Test ModelConnectionTransformer with keyField overrides the default limit', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(limit: 50, fields: ["id"])
    }
    type Comment @model {
        id: ID!
        content: String
    }
    `;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });

  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();

  // Post.comments field
  expect(out.resolvers['Post.comments.req.vtl']).toContain('#set( $limit = $util.defaultIfNull($context.args.limit, 50) )');
});

test('Test ModelConnectionTransformer with keyField uses the default limit', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        title: String!
        comments: [Comment] @connection(fields: ["id"])
    }
    type Comment @model {
        id: ID!
        content: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.ConnectionStack.Resources[ResolverResourceIDs.ResolverResourceID('Post', 'comments')]).toBeTruthy();

  // Post.comments field
  expect(out.resolvers['Post.comments.req.vtl']).toContain(
    `#set( $limit = $util.defaultIfNull($context.args.limit, ${ResourceConstants.DEFAULT_PAGE_LIMIT}) )`,
  );
});

test('Connection on models with no codegen includes AttributeTypeEnum', () => {
  const validSchema = `
    type Post @model(queries: null, mutations: null, subscriptions: null) {
        id: ID!
        title: String!
        comments: [Comment] @connection
    }
    type Comment @model(queries: null, mutations: null, subscriptions: null) {
        id: ID!
        content: String
    }
  `;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
    transformConfig: {
      Version: TRANSFORM_CURRENT_VERSION,
    },
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();

  expect(out.schema).toMatchSnapshot();
});

test('Connection on models with no codegen includes custom enum filters', () => {
  const validSchema = `
    type Cart @model(queries: null, mutations: null, subscriptions: null) {
      id: ID!,
      cartItems: [CartItem] @connection(name: "CartCartItem")
    }
    
    type CartItem @model(queries: null, mutations: null, subscriptions: null) {
      id: ID!
      productType: PRODUCT_TYPE!
      cart: Cart @connection(name: "CartCartItem")
    }
    
    enum PRODUCT_TYPE {
      UNIT
      PACKAGE
    }
  `;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new ModelConnectionTransformer()],
    transformConfig: {
      Version: TRANSFORM_CURRENT_VERSION,
    },
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toMatchSnapshot();
});

function expectFields(type: ObjectTypeDefinitionNode, fields: string[]) {
  for (const fieldName of fields) {
    const foundField = type.fields.find((f: FieldDefinitionNode) => f.name.value === fieldName);
    expect(foundField).toBeDefined();
  }
}

function expectArguments(field: FieldDefinitionNode, args: string[]) {
  for (const argName of args) {
    const foundArg = field.arguments.find((a: InputValueDefinitionNode) => a.name.value === argName);
    expect(foundArg).toBeDefined();
  }
}

function doNotExpectFields(type: ObjectTypeDefinitionNode, fields: string[]) {
  for (const fieldName of fields) {
    expect(type.fields.find((f: FieldDefinitionNode) => f.name.value === fieldName)).toBeUndefined();
  }
}

function getObjectType(doc: DocumentNode, type: string): ObjectTypeDefinitionNode | undefined {
  return doc.definitions.find((def: DefinitionNode) => def.kind === Kind.OBJECT_TYPE_DEFINITION && def.name.value === type) as
    | ObjectTypeDefinitionNode
    | undefined;
}

function getInputType(doc: DocumentNode, type: string): InputObjectTypeDefinitionNode | undefined {
  return doc.definitions.find((def: DefinitionNode) => def.kind === Kind.INPUT_OBJECT_TYPE_DEFINITION && def.name.value === type) as
    | InputObjectTypeDefinitionNode
    | undefined;
}

function verifyInputCount(doc: DocumentNode, type: string, count: number): boolean {
  return doc.definitions.filter(def => def.kind === Kind.INPUT_OBJECT_TYPE_DEFINITION && def.name.value === type).length === count;
}
