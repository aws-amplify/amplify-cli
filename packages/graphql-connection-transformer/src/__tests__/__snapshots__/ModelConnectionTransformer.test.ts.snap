// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection on models with no codegen includes AttributeTypeEnum 1`] = `
"type Post {
  id: ID!
  title: String!
  comments(filter: ModelCommentFilterInput, sortDirection: ModelSortDirection, limit: Int, nextToken: String): ModelCommentConnection
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

type Comment {
  id: ID!
  content: String
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

type ModelCommentConnection {
  items: [Comment]!
  nextToken: String
}

enum ModelSortDirection {
  ASC
  DESC
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

input ModelCommentFilterInput {
  id: ModelIDInput
  content: ModelStringInput
  and: [ModelCommentFilterInput]
  or: [ModelCommentFilterInput]
  not: ModelCommentFilterInput
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}
"
`;

exports[`Connection on models with no codegen includes custom enum filters 1`] = `
"type Cart {
  id: ID!
  cartItems(filter: ModelCartItemFilterInput, sortDirection: ModelSortDirection, limit: Int, nextToken: String): ModelCartItemConnection
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

type CartItem {
  id: ID!
  productType: PRODUCT_TYPE!
  cart: Cart
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

enum PRODUCT_TYPE {
  UNIT
  PACKAGE
}

type ModelCartItemConnection {
  items: [CartItem]!
  nextToken: String
}

enum ModelSortDirection {
  ASC
  DESC
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

input ModelPRODUCT_TYPEInput {
  eq: PRODUCT_TYPE
  ne: PRODUCT_TYPE
}

input ModelCartItemFilterInput {
  id: ModelIDInput
  productType: ModelPRODUCT_TYPEInput
  and: [ModelCartItemFilterInput]
  or: [ModelCartItemFilterInput]
  not: ModelCartItemFilterInput
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}
"
`;
